### DISCUSSION TITLE
How does bluetooth work while the app is backgrounded?

### DISCUSSION THREAD

**Discussion by aduckworth11** [capacitor-community/bluetooth-le#679] [Q&A]
I am attempting to scan for a specific bluetooth device while the app is in the foreground, but also while the app is in the background (and even with the phone locked). On iOS I have added bluetooth-central to UIBackgroundModes, however the code in the callback for requestLEScan never seems to trigger when the app is backgrounded. I'm not sure if the scanning is not working, or it is unable to run the hybrid code while the app is backgrounded. 

Android seems to be working as intended when passing the scan mode of SCAN_MODE_LOW_POWER.

Is this not the way background bluetooth scanning works with this plugin? 

Any help would be appreciated. Thank you!

**Comment by peitschie** [comment]
Hi @aduckworth11 

On iOS, background scanning is a bit finicky. How long are you pushing the app into the background and waiting before you'd expect the scan response to trigger?

With the default approach used by this plugin, the app will not be correctly reloaded if iOS pauses the app in the background, as there is no support for iOS state restoration. If you had a simple enough test harness... it would be interesting to know if another plugin (https://github.com/don/cordova-plugin-ble-central?tab=readme-ov-file#background-scanning-and-notifications-on-ios) might behave differently here?


Just as an aside, with the Android app you'll need to wait **at least** 5 minutes to properly test that your application responds. Most hybrid web views pause execution of JS code 5 minutes after being put into the background. Additionally, Android power saving mode tends to kick in about 40mins => 1hr later if the app remains in the background.

**Reply by aduckworth11** [reply]
Hi @peitschie 

There wasn't really a specific time frame I was looking for, just any scanning at all. After I close the app (move to the background, not fully exit the app) I can't really seem to get any scanning at all. Once the app is paused by iOS I wouldn't expect the BLE background scanning to fully still work, and I know that can take a variable amount of time, although that is interesting the cordova version seems to support that with state restoration.

I just tried my same implementation using the cordova version of the plugin with no luck either. I made sure to turn on the bluetooth_restore_state option as well.

I'm essentially running a test of pushing some data I get from my BLE device with a timestamp to an array whenever the device is found and showing this list of scans on the screen. This works fine with the app in the foreground on iOS, but when backgrounded it doesn't seem to scan. At least I am deriving this from the time stamp I have associated with each scan.

Ex. I scan the device at 4:05 PM with the app in the foreground. Background the app for X amount of time (could be 5 minutes could be an hour). Open the app and I'm not seeing any scans with javascript timestamps in between the time I backgrounded the app and re-opened it. I may get a new scan shortly after bringing the app to the foreground but that isn't a background scan (at least not from what I can tell). On android I was able to get pretty consistent background scans whether it would be 5 minutes or 30. I was seeing timestamps associated with the data on the screen from times where I had the app backgrounded.

Would it be possible that this is not a reliable test? Maybe the background scan is cached and the success callback triggers when the app re-opens causing the timestamp to not be an accurate reflection of when the scan actually happens?

**Reply by peitschie** [reply]
Are you setting `reportDuplicates` to true? Are you using any service UUID filters in your scan?

Apologies, I don't really have a clear problem to point you at. I suspect it's not the plugin, as I have definitely used iOS background scanning successfully before, but there are a lot of little things that need to get aligned for this to function reliably.

Though it may not help at all, it might be worth reading through https://medium.com/@cbartel/ios-scan-and-connect-to-a-ble-peripheral-in-the-background-731f960d520d and seeing if there are any hints here that might be relevant?

**Reply by aduckworth11** [reply]
Ah it was the services filter I was not using! I have now been able to get several successful background scans in iOS using both the capacitor and bluetooth plugins. Thank you so much for your help and apologies for any inconvenience!

**Reply by peitschie** [reply]
That's fantastic to hear! Glad you were able to figure this out.
